### ISSUE TITLE
can't run falcon with websocket - undefined method `call' for Falcon::Adapters::Output

### ISSUE THREAD

**Issue by senid231** [socketry/falcon#93]
NoMethodError: undefined method `call' for #<Falcon::Adapters::Output length=nil body=Rack::BodyProxy>

## Reproduce

use following project

https://github.com/senid231/falcon-test

or create your own with following files

config.ru
```ruby
require 'logger'
require 'async/websocket'
require 'async/websocket/adapters/rack'

APP_LOGGER = Logger.new(STDOUT)

module AsyncCable
  class Connection < Async::WebSocket::Connection
    def handle_command(message)
      APP_LOGGER.info { "AsyncCable::Connection#handle_command message=#{message.inspect}" }
    end

    def handle_open
      APP_LOGGER.info { "AsyncCable::Connection#handle_open." }
    end

    def handle_close(clean, code = nil)
      APP_LOGGER.info { "AsyncCable::Connection#handle_close clean=#{clean} code=#{code}" }
      close
    end
  end

  class Server
    def initialize(connection_class:)
      @connection_class = connection_class
    end

    def call(env)
      Async::WebSocket::Adapters::Rack.open(env, handler: @connection_class) do |connection|
        connection.handle_open

        while (data = connection.read)
          connection.handle_command(data)
        end
      rescue Protocol::WebSocket::ClosedError => error
        APP_LOGGER.info { "AsyncCable::Server ClosedError error=#{error.message}" }
        connection.handle_close(false, error.code)
      ensure
        APP_LOGGER.info { "AsyncCable::Server closed." }
        connection.handle_close(true, Protocol::WebSocket::Connection::Error::NO_ERROR)
      end or [200, {}, ['Hello World']]
    end
  end
end

app = Rack::Builder.new do
  map '/' do
    use Rack::CommonLogger, APP_LOGGER
    use Rack::Static, urls: %w(/assets /index.html), root: File.join(__dir__, 'public')
    run proc { |_| raise StandardError, 'no file' }
  end

  map '/cable' do
    use Rack::CommonLogger, APP_LOGGER
    run AsyncCable::Server.new(connection_class: AsyncCable::Connection)
  end
end

run app
```

public/index.html
```html
<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width">
  <title>Test Falcon with websocket</title>
<body>

<script type="text/javascript" src="/assets/cable.js"></script>

</body>
</html>
```

public/assets/cable.js
```js
var initializeSocket = function () {
  var socket = new WebSocket("ws://localhost:4567/cable");


  socket.onopen = function(_event) {
    console.log("WebSocket connected");
  };

  socket.onclose = function(event) {
    if (event.wasClean) {
      console.log("WebSocket connection closed");
    } else {
      console.log("WebSocket connection closed dirty"); // for example server process was killed
    }
    console.log('WebSocket code: ' + event.code + ' reason: ' + event.reason);
  };

  socket.onmessage = function(event) {
    console.log("WebSocket data received " + event.data);
  };

  socket.onerror = function(error) {
    console.log("WebSocket error " + error.message);
  };

  return socket
}

window.MainApp = { socket: null };
MainApp.socket = initializeSocket();
```

Gemfile
```ruby
source 'https://rubygems.org'
git_source(:github) { |repo| "https://github.com/#{repo}.git" }

ruby '2.6.0'

gem 'activesupport'

# Async web server
# https://github.com/socketry/falcon
gem 'falcon'

# Async websocket implementation
# https://github.com/socketry/async-websocket
gem 'async-websocket'

group :development do
  gem 'byebug'
end
```

Gemfile.lock
```
GEM
  remote: https://rubygems.org/
  specs:
    activesupport (6.0.2.1)
      concurrent-ruby (~> 1.0, >= 1.0.2)
      i18n (>= 0.7, < 2)
      minitest (~> 5.1)
      tzinfo (~> 1.1)
      zeitwerk (~> 2.2)
    async (1.24.0)
      console (~> 1.0)
      nio4r (~> 2.3)
      timers (~> 4.1)
    async-container (0.14.1)
      async (~> 1.0)
      async-io (~> 1.4)
      process-group
    async-http (0.49.1)
      async (~> 1.23)
      async-io (~> 1.27.0)
      protocol-http (~> 0.13.0)
      protocol-http1 (~> 0.10.0)
      protocol-http2 (~> 0.10.0)
    async-io (1.27.1)
      async (~> 1.14)
    async-websocket (0.13.1)
      async-http (~> 0.43)
      async-io (~> 1.23)
      protocol-websocket (~> 0.7.0)
    build-environment (1.12.1)
    byebug (11.0.1)
    concurrent-ruby (1.1.5)
    console (1.6.0)
    falcon (0.34.4)
      async (~> 1.13)
      async-container (~> 0.14.0)
      async-http (~> 0.49.0)
      async-io (~> 1.22)
      build-environment (~> 1.11)
      localhost (~> 1.1)
      rack (>= 1.0)
      samovar (~> 2.1)
    ffi (1.11.3)
    i18n (1.7.0)
      concurrent-ruby (~> 1.0)
    localhost (1.1.6)
    mapping (1.1.1)
    minitest (5.13.0)
    nio4r (2.5.2)
    process-group (1.2.1)
      process-terminal (~> 0.2.0)
    process-terminal (0.2.0)
      ffi
    protocol-hpack (1.4.1)
    protocol-http (0.13.0)
    protocol-http1 (0.10.0)
      protocol-http (~> 0.13)
    protocol-http2 (0.10.4)
      protocol-hpack (~> 1.4)
      protocol-http (~> 0.2)
    protocol-websocket (0.7.4)
      protocol-http (~> 0.2)
      protocol-http1 (~> 0.2)
    rack (2.0.8)
    samovar (2.1.2)
      console (~> 1.0)
      mapping (~> 1.0)
    thread_safe (0.3.6)
    timers (4.3.0)
    tzinfo (1.2.5)
      thread_safe (~> 0.1)
    zeitwerk (2.2.2)

PLATFORMS
  ruby

DEPENDENCIES
  activesupport
  async-websocket
  byebug
  falcon

RUBY VERSION
   ruby 2.6.0p0

BUNDLED WITH
   2.1.2
```

`RACK_ENV=development bundle exec bin/falcon serve -b http://localhost:4567 -n 1 --threaded`

open http://localhost:4567/index.html in browser

server log
```
 0.12s     info: #<Falcon::Endpoint http://localhost:4567/ {}> [pid=18448] [2019-12-21 12:51:28 +0200]
               | Falcon v0.34.4 taking flight! Using Async::Container::Threaded {:count=>1}
               | - To terminate: Ctrl-C or kill 18448
 0.12s     info: #<Async::Container::Threaded::Instance:0x00005648e093dfa8> [pid=18448] [2019-12-21 12:51:28 +0200]
               | - Per-process status: kill -USR1 18448
::1 - - [21/Dec/2019:12:51:49 +0200] "GET /index.html " 304 - 17.5442
::1 - - [21/Dec/2019:12:51:51 +0200] "GET /assets/cable.js " 304 - 1.4905
34.85s    error: #<Falcon::Adapters::Rack:0x00005648e0c653c0> [pid=18448] [2019-12-21 12:52:03 +0200]
               |   StandardError: no file
               |   → config.ru:50 in `block (4 levels) in <main>'
               |     /home/senid/.rvm/gems/ruby-2.6.0@falcon-test/gems/rack-2.0.8/lib/rack/static.rb:149 in `call'
               |     /home/senid/.rvm/gems/ruby-2.6.0@falcon-test/gems/rack-2.0.8/lib/rack/common_logger.rb:33 in `call'
               |     /home/senid/.rvm/gems/ruby-2.6.0@falcon-test/gems/rack-2.0.8/lib/rack/urlmap.rb:68 in `block in call'
               |     /home/senid/.rvm/gems/ruby-2.6.0@falcon-test/gems/rack-2.0.8/lib/rack/urlmap.rb:53 in `each'
               |     /home/senid/.rvm/gems/ruby-2.6.0@falcon-test/gems/rack-2.0.8/lib/rack/urlmap.rb:53 in `call'
               |     /home/senid/.rvm/gems/ruby-2.6.0@falcon-test/gems/rack-2.0.8/lib/rack/builder.rb:153 in `call'
               |     /home/senid/.rvm/gems/ruby-2.6.0@falcon-test/gems/falcon-0.34.4/lib/falcon/adapters/rack.rb:181 in `call'
               |     /home/senid/.rvm/gems/ruby-2.6.0@falcon-test/gems/protocol-http-0.13.0/lib/protocol/http/middleware.rb:40 in `call'
               |     /home/senid/.rvm/gems/ruby-2.6.0@falcon-test/gems/falcon-0.34.4/lib/falcon/adapters/rewindable.rb:61 in `call'
               |     /home/senid/.rvm/gems/ruby-2.6.0@falcon-test/gems/protocol-http-0.13.0/lib/protocol/http/middleware.rb:40 in `call'
               |     /home/senid/.rvm/gems/ruby-2.6.0@falcon-test/gems/protocol-http-0.13.0/lib/protocol/http/content_encoding.rb:44 in `call'
               |     /home/senid/.rvm/gems/ruby-2.6.0@falcon-test/gems/protocol-http-0.13.0/lib/protocol/http/middleware.rb:40 in `call'
               |     /home/senid/.rvm/gems/ruby-2.6.0@falcon-test/gems/async-http-0.49.1/lib/async/http/server.rb:64 in `block in accept'
               |     /home/senid/.rvm/gems/ruby-2.6.0@falcon-test/gems/async-http-0.49.1/lib/async/http/protocol/http1/server.rb:58 in `each'
               |     /home/senid/.rvm/gems/ruby-2.6.0@falcon-test/gems/async-http-0.49.1/lib/async/http/server.rb:53 in `accept'
               |     /home/senid/.rvm/gems/ruby-2.6.0@falcon-test/gems/async-io-1.27.1/lib/async/io/socket.rb:99 in `block in accept_each'
               |     /home/senid/.rvm/gems/ruby-2.6.0@falcon-test/gems/async-io-1.27.1/lib/async/io/socket.rb:137 in `block in accept'
               |     /home/senid/.rvm/gems/ruby-2.6.0@falcon-test/gems/async-1.24.0/lib/async/task.rb:256 in `block in make_fiber'
::1 - - [21/Dec/2019:12:52:18 +0200] "GET /index.html " 304 - 0.0002
50.17s    error: #<Async::Task:0x2b24705f7dcc incoming connection #<Addrinfo: [::1]:57524 TCP> [fd=17] (failed)> [pid=18448] [2019-12-21 12:52:18 +0200]
               |   NoMethodError: undefined method `call' for #<Falcon::Adapters::Output length=nil body=Rack::BodyProxy>
               |   → /home/senid/.rvm/gems/ruby-2.6.0@falcon-test/gems/async-http-0.49.1/lib/async/http/protocol/http1/server.rb:76 in `each'
               |     /home/senid/.rvm/gems/ruby-2.6.0@falcon-test/gems/async-http-0.49.1/lib/async/http/server.rb:53 in `accept'
               |     /home/senid/.rvm/gems/ruby-2.6.0@falcon-test/gems/async-io-1.27.1/lib/async/io/socket.rb:99 in `block in accept_each'
               |     /home/senid/.rvm/gems/ruby-2.6.0@falcon-test/gems/async-io-1.27.1/lib/async/io/socket.rb:137 in `block in accept'
               |     /home/senid/.rvm/gems/ruby-2.6.0@falcon-test/gems/async-1.24.0/lib/async/task.rb:256 in `block in make_fiber'
```

browser log
```
WebSocket connected
WebSocket connection closed dirty
WebSocket code: 1006 reason: 
```

**Comment by senid231** [comment]
I've fixed in by this patch:
```ruby
Falcon::Adapters::Output.class_eval do
  def call(stream)
    @body.call(stream)
  end
end
```

**Comment by senid231** [comment]
btw this solution not ideal
when using `Rack::Protection::CookieTossing` it breaks because in this case `@body` is `Rack::BodyProxy` for `Rack::Response` (instead of `Async::HTTP::Body::Hijack`)

**Comment by ioquatix** [comment]
Can you show me where in the source code `Rack::Protection::CookieTossing` turns the response into a `BodyProxy`?

**Comment by senid231** [comment]
@ioquatix here's what happening

`rack-protection-2.0.7/lib/rack/protection/cookie_tossing.rb:22 in #call`
```ruby
response = Rack::Response.new(body, status, headers)
```
`rack-protection-2.0.7/lib/rack/protection/cookie_tossing.rb:25 in #call`
```ruby
response.finish
```
`rack-2.0.8/lib/rack/response.rb:69 in #finish`
```ruby
[status.to_i, header, BodyProxy.new(self){}]
```
`Rack::Protection::CookieTossing` wraps response into `Rack::Response` which reads all from body and return `Rack::BodyProxy`as last element of response array.

**Comment by ioquatix** [comment]
`Rack::Response` is poorly designed and turns any body into a body proxy, even if it was unnecessary. I'm attempting to fix this.

**Comment by ioquatix** [comment]
This should be fixed by Rack v2.1.0.
